### PR TITLE
[semver] Stricter `options` types

### PR DESCRIPTION
**Pull request by aried3r**
Only functions that take ranges take a `includePrereleases` option.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
  The last paragraph [here](https://github.com/npm/node-semver/tree/v7.3.7#prerelease-tags) contains the following:
  > by setting the `includePrerelease` flag on the options object to any [functions](https://github.com/npm/node-semver#functions) **that do range matching**.

  Functions can be found here: https://github.com/npm/node-semver/tree/v7.3.7/functions
  Tests and their fixtures can be found here: https://github.com/npm/node-semver/tree/v7.3.7/test/functions
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

File: types/semver/index.d.ts [hunk 1]
```diff
@@ -113,6 +113,9 @@ export type ReleaseType = 'major' | 'premajor' | 'minor' | 'preminor' | 'patch'
 
 export interface Options {
     loose?: boolean | undefined;
+}
+
+export interface RangeOptions extends Options {
     includePrerelease?: boolean | undefined;
 }
```
Comment (aried3r) [inline types/semver/index.d.ts:120]: This is a breaking change if people use `Semver.Options` right now, but I couldn't find a better name. There are no "general" options in the `semver` package. Open for suggestions!

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## semver ([<kbd>unpkg</kbd>](https://unpkg.com/browse/semver@latest/))
was missing the following properties:
1. re
1. src
1. tokens
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against ab41314ba4d4ddf63296dff911136c190043633b
</p>

File: types/semver/index.d.ts (hunk 1 above)
Comment (peterblazejewicz) [reply types/semver/index.d.ts:120]: aren't all options supposed to support this prop? Say:
https://github.com/npm/node-semver/blob/802e16174fe2a704dba16e97891ce36dc4f2ad76/classes/semver.js#L33
('false' by default, can be boolean or an object with `loose` or `includePrerelease`)

Comment (aried3r) [reply types/semver/index.d.ts:120]: Hey @peterblazejewicz! Thanks for your question.

As for the code you linked, you can see the [comment above](https://github.com/npm/node-semver/blob/802e16174fe2a704dba16e97891ce36dc4f2ad76/classes/semver.js#L31-L33), saying:

> this isn't actually relevant for versions

While the constructor takes the additional argument (as it does in this PR), if you import the functions directly, they are not needed, omitted and have no effect. By being more strict with the types, we can avoid pitfalls.

For example, let's take the `compare` function:

https://github.com/npm/node-semver/blob/802e16174fe2a704dba16e97891ce36dc4f2ad76/functions/compare.js

It uses the constructor, but is only interested in `loose`.

When checking the tests, we can also only see that it's being tested with a `boolean` and an object: `{ loose: loose }`:

https://github.com/npm/node-semver/blob/802e16174fe2a704dba16e97891ce36dc4f2ad76/test/functions/compare.js

The fixtures for the test also only ever provide a `loose` argument:

https://github.com/npm/node-semver/blob/802e16174fe2a704dba16e97891ce36dc4f2ad76/test/fixtures/comparisons.js

And as their README says:

> setting the `includePrerelease` flag on the options object to any [functions](https://github.com/npm/node-semver/tree/v7.3.7#functions) that **do range matching**.

I'll take `satisfies` as an example, linking only the fixtures in this case. It takes no options, a `boolean`, and an object, namely what I called `RangeOptions`:

https://github.com/npm/node-semver/blob/v7.3.7/test/fixtures/range-exclude.js
https://github.com/npm/node-semver/blob/v7.3.7/test/fixtures/range-include.js

Does that answer your question?

Comment (rbuckton) [reply types/semver/index.d.ts:120]: This makes sense to me.

**Review by rbuckton — APPROVED** [review]

**Comment by aried3r** [comment]
Ready to merge

**Comment by nick4598** [comment]
Shouldn't all functions which take in a range have also been updated to reflect that they can take in semver.RangeOptions and not just semver.Options?

I'm specifically running into an issue with maxSatisfying which does take in a range, but the types for it are "optionsOrLoose?: boolean | semver.Options," which results in a buildtime error. 

" Argument of type '{ loose: true; includePrerelease: boolean | undefined; }' is not assignable to parameter of type 'boolean | Options | undefined'.
  Object literal may only specify known properties, and 'includePrerelease' does not exist in type 'Options'"

**Comment by aried3r** [comment]
Great catch, @nick4598! A potential fix is in #61658.

**Comment by peterblazejewicz** [comment]
Should be fixed in next version

**Comment by mrgrain** [comment]
We are running into the same issue with `parse()` now:

```
Error: src/util/semver.ts:93:44 - error TS2345: Argument of type '{ includePrerelease: boolean; }' is not assignable to parameter of type 'boolean | Options | undefined'.
  Object literal may only specify known properties, and 'includePrerelease' does not exist in type 'Options'.

[9](https://github.com/projen/projen/runs/7803995052?check_suite_focus=true#step:5:10)3   const version = parse(assemblyVersion, { includePrerelease: true });
```

From the source I can see that `parse(version, options)` is a simple wrapper for `new SemVer(version, options)` which does take a range. See https://github.com/npm/node-semver/blob/main/functions/parse.js#L27

**Comment by aried3r** [comment]
@mrgrain, there's #61658 which fixes some issues. As for `parse()`, I don't think it does anything with `{ includePrereleases: true }`, at least it's not tested and it does not handle ranges.

https://github.com/npm/node-semver/blob/v7.3.7/test/functions/parse.js#L4
https://github.com/npm/node-semver/blob/v7.3.7/test/fixtures/invalid-versions.js

```JS
semver.parse('^1.0.0') // null
```

**Comment by mrgrain** [comment]
Thanks for looking into this @aried3r. I've dug around the `semver` code base a bit more and found this:
https://github.com/npm/node-semver/blob/e9a1f26e616b2e7a3e14e1803364229980142789/classes/semver.js#L31-L33

I agree that semantically this doesn't make sense. Before this change, it was possible to use `parse` to create a `SemVer` object with `SemVer.includePrerelease = true`. That's gone now if anyone relied on this. I'm not familiar what the goals and rules around DefinitelyTyped are, so leaving this entirely up to you. 😀 

My test suite passes without the option, so I'm okay either way.

**Comment by peterblazejewicz** [comment]
maybe it was there, as consuming code is not always aware which object type was passed around, say when constructed 'parse' was re-exported and imported in other case (think service module), so the strict types resolution would break this behavior.

File: types/semver/index.d.ts (hunk 1 above)
Comment (edzis) [reply types/semver/index.d.ts:120]: Removing `includePrerelease` from the generic Options seems to go against the documentation - https://github.com/npm/node-semver/blob/main/README.md?plain=1#L364-L378

>All methods and classes take a final `options` object argument.  All
options in this object are `false` by default.  The options supported
are:
> - `loose`  ...
> - `includePrerelease`  ...

Would you say that the documentation is misleading about the actual effect of the `includePrerelease` flag?

**Pull request merged**